watchtower/wtdb/tower_db.go

```diff
@@ -0,0 +1,577 @@
+	"github.com/lightningnetwork/lnd/chainntnfs"
+	"github.com/lightningnetwork/lnd/channeldb"
+)
+
+const (
+	// dbName is the filename of tower database.
+	dbName = "watchtower.db"
+
+	// dbFilePermission requests read+write access to the db file.
+	dbFilePermission = 0600
+)
+
+var (
+	// sessionsBkt is a bucket containing all negotiated client sessions.
+	//  session id -> session
+	sessionsBkt = []byte("sessions-bucket")
+
+	// updatesBkt is a bucket containing all state updates sent by clients.
+	// The updates are further bucketed by session id to prevent clients
+	// from overwrite each other.
+	//   hint => session id -> update
+	updatesBkt = []byte("updates-bucket")
+
+	// lookoutTipBkt is a bucket containing the last block epoch processed
+	// by the lookout subsystem. It has one key, lookoutTipKey.
+	//   lookoutTipKey -> block epoch
+	lookoutTipBkt = []byte("lookout-tip-bucket")
+
+	// lookoutTipKey is a static key used to retrieve lookout tip's block
+	// epoch from the lookoutTipBkt.
+	lookoutTipKey = []byte("lookout-tip")
+
+	// metadataBkt stores all the meta information concerning the state of
+	// the database.
+	metadataBkt = []byte("metadata-bucket")
+
+	// dbVersionKey is a static key used to retrieve the database version
+	// number from the metadataBkt.
+	dbVersionKey = []byte("version")
+
+	// ErrUninitializedDB signals that top-level buckets for the database
+	// have not been initialized.
+	ErrUninitializedDB = errors.New("tower db not initialized")
+
+	// ErrNoDBVersion signals that the database contains no version info.
+	ErrNoDBVersion = errors.New("tower db has no version")
+
+	byteOrder = binary.BigEndian
+)
+
+// TowerDB is single database providing a persistent storage engine for the
+// wtserver and lookout subsystems.
+type TowerDB struct {
+	db     *bbolt.DB
+	dbPath string
+}
+
+// OpenTowerDB opens the tower database given the path to the database's
+// directory. If no such database exists, this method will initialize a fresh
+// one using the latest version number and bucket structure. If a database
+// exists but has a lower version number than the current version, any necessary
+// migrations will be applied before returning. Any attempt to open a database
+// with a version number higher that the latest version will fail to prevent
+// accidental reversion.
+func OpenTowerDB(dbPath string) (*TowerDB, error) {
+	path := filepath.Join(dbPath, dbName)
+
+	// If the database file doesn't exist, this indicates we much initialize
+	// a fresh database with the latest version.
+	firstInit := !fileExists(path)
+	if firstInit {
+		// Ensure all parent directories are initialized.
+		err := os.MkdirAll(dbPath, 0700)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	bdb, err := bbolt.Open(path, dbFilePermission, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	towerDB := &TowerDB{
+		db:     bdb,
+		dbPath: dbPath,
+	}
+
+	if firstInit {
+		// The file didn't exist before, initialize a fresh database
```

Comment:
Seems possible for the file to have been created, but the bucket structure left blank. Therefore, we should just check the bucket directly.

Reply:
wdym by check the bucket directly?

I reworked the logic slightly in e5c34751bad0efd0fd9448a0f59946530fc739b7  to handle setting/migration versions first, then initializing all top-level buckets

Reply:
I mean just always call the init logic as if the bucket exists, this it's a noop.